typeclass-experiments/src/Gen/Generator.hs

```diff
@@ -0,0 +1,68 @@
+module Gen.Generator where
+
+import Control.Monad.Trans.Class
+import Control.Monad.Trans.State
+import Data.Set (Set)
+import Data.Set qualified as S
+import Data.Text (Text)
+import Data.Text.IO qualified as T
+
+import Common.Match
+import Common.SourceTy
+import Common.Types
+
+data GenError
+  = MatchFail Pat Pat
+  | DeriveFail Pat
+  | MultipleMatches Pat [Pat]
+  | MalformedPatList Pat
+  deriving (Show)
+
+type GenResult = Either GenError Text
+
+-- crude code generator
+data Gen = Gen
+  { genPat :: Pat
+  , runGen :: Set Gen -> Pat -> GenResult
+  }
```

Comment:
This could be called `TypeDefPrinter`? As in, this just maps type definitions to Text? Just trying to form a mental map of this.

As mentioned elsewhere, let's not map directly to Text as it's very unmaintainable and overly complex, instead formulate a concise Rust DSL so we can talk in terms of RustModule, RustTypeDeclaration, RustFunction, RustBultinXYZ, RustStatement etc. And afterwards we can have a super simple Text printing facility implemented.

Reply:
It's a way of bundling modular code generation functions with a first class representation of the pattern that they generate code for. This _is a framework for writing a DSL like that_. Like, here, example: 

``` haskell 
rustList :: Gen
rustList = Gen (List _x) $ \scope pat -> case pat of
  List t -> do
    t' <- runGenerator scope t
    pure $ "Vec<" <> t' <> ">"
  other -> matchFail (List _x) other
```

`List _x` is just sugar for `AppP ListP (VarP "x")` or, if you really wanna do things with no builtins, `AppP (RefP "List") (VarP "x")`. What that means is: "I'll match with any [a], no matter what [a] is". That's the part that determines which pattern the generator is a generator for. 

The function part takes a `Set Gen` (`scope` for lack of a better name) and a `Pat`. 

*We* know that the `Pat` is a substitution instance of `List _x / AppP ListP (VarP "x")` but we still have to match on it to get ahold of the Haskell variable `t` (in the Haskell AST it's a `Language.Haskell.TH.Syntax VarP`) which corresponds to the variable in our pattern. 

(Given enough time I could write some template haskell that converts our Pat into a TH Pat, or use fancy type stuff to enforce the structural similarity of the `Pat` in the function  to the `Pat` arg, but anyway let's pretend I won't do that for now since you hate that stuff. The _annoying_ part this about this is that we know that the `other` branch is inaccessible here but without fanciness we can't tell Haskell that it is)

The only thing we can do with `t` is try to generate some code for it, since `rustList` only knows how to wrap an argument type with `Vec<_>`. So we do that.

Reply:
This is _kind of_ like a combination of a parser and pretty printer. The `genPat :: Pat` field in `Gen` tells you what to match (the parser), and the function part tells you what to construct from the match (the pretty printer). It *could* be 

``` haskell 
-- e is an error type, a is the result type 
data Gen e a = Gen
  { genPat :: Pat
  , runGen :: Set (Gen e a) -> Pat -> Either e a
  }
```

In principle though I don't think there would be a reason to do `Gen GenError RustDSL` instead of `Gen GenError ([Import],Doc)` though.

Reply:
I'm liking this! I do still think we'd benefit from Rust DSL and adopting that approach throughout